### ISSUE TITLE
Heatingcircuit plugin: fatal KeyError with sensor values (hctemp, outside_temp)

### ISSUE THREAD

**Issue by tsoukase** [motoz/PellMon#120]
When the Heatingcircuit plugin is activated and pellmonsrv is started,  I get:

<pre>Traceback (most recent call last):
  File "/usr/local/bin/pellmonsrv", line 22, in <module>
    pellmonsrv.run()
  File "/usr/local/lib/python2.7/dist-packages/Pellmonsrv/pellmonsrv.py", line 850, in run
    commands[args.command]()
  File "/usr/local/lib/python2.7/dist-packages/Pellmonsrv/pellmonsrv.py", line 437, in run
    conf.database = Database()
  File "/usr/local/lib/python2.7/dist-packages/Pellmonsrv/pellmonsrv.py", line 89, in __init__
    plugin.plugin_object.activate(conf.plugin_conf[plugin.name], globals(), self)
  File "/usr/local/lib/python2.7/dist-packages/Pellmonsrv/plugins/heatingcircuit/__init__.py", line 63, in activate
    self.desired_temp = float(self.db[self.hctemp_item].value)
  File "/usr/lib/python2.7/weakref.py", line 83, in __getitem__
    o = self.data[key]()
KeyError: 'T7'
No option 'plugin_dirs' in section: 'plugin_settings'
</pre>

and pellmonsrv exits.

My heatingcircuit.conf, based on the connection of my 3-way valve, is: 

<pre>output_close = L10 # no problem
output_open = L9 # no problem
hctemp = T7 # THIS causes the problem
outside_temp = T5
</pre>

I tried inserting 7/5 and lng_T5/lng_T5 with the same problem.

If you can help with this issue, I would appreciate it.

**Comment by motoz** [comment]
I'll at least try! (I have been using the heatingcircuit plugin myself for many years so it should work...). First, if you disable heatingcircut so you can get it to start, do you then have an item called T7 with a valid temperature reading? If so, from which plugin does that come? If I remember correctly (couple of years since I touched this) you need to list that plugin before heatingcircuit in enabled_plugins.conf so the item is already created when heatingcircuit is initialized.

**Comment by tsoukase** [comment]
Thanks for your interest.
At page "weather" it says:
<pre>
Input- forwarding temp. sensor    7 
Input-outside temp. sensor     5 
Regulating valve close     L10
Outlet house pump     OFF 
Regulating valve open    L9
</pre>

with the inputs having "Valid values: 0 - 20". As said, I have tried 7/5, too.
I find their corresponding values at page "operating_data", named forward_ref and forward_temp.
I use the provided "enabled_plugins.conf", with
<pre>

p07 = PelletCalc
p08 = Consumption
p11 = Heatingcircuit
p12 = NBEcom
</pre>

I tried changing the order and the numbers p11/12 with no success.

**Comment by motoz** [comment]
Oh, I think you have mixed things up a bit. If I understand you correctly you are using an NBE burner controller to regulate your 3-way valve. The heatingcircuit plugin is for those that doesn't have that functionality build in (or for some reason doesn't want to use it). 

I have two gpio pins on the raspberry that runs pellmon mapped as outputs using the raspberryGPIO plugin like this:
gpio2_function = output
gpio2_item = **out_close**
gpio2_pin = 11
gpio3_function = output
gpio3_item = **out_open**
gpio3_pin = 13

These two outputs are named 'out_close' and 'out_open'. They are connected to the heatingcircuit plugin like this:
output_close = **out_close**
output_open = **out_open**
hctemp = heating_circuit
outside_temp = outside_temp

The 'heating_circuit' item comes from the OWFS plugin, and contains the temperature measured using a onewire sensor. 'outside_temp' is an item containing the coldest value of onewire sensors on the north and south side of my house, calculated by the 'pelletcalc' plugin using this conf:
c2_prog = outside_n_temp get outside_s_temp get min
c2_readitem = **outside_temp**

I then have some custom electronics connecting the raspberry gpio outputs to an old danfoss 24V motor that turns the valve.  All this is completely independent from the NBE burner controller.

tl;dr. You don't need the heatingcircuit plugin together with an NBE burner, but you can use it if you really want to...

**Comment by tsoukase** [comment]
You are right. My NBE v7 Controller drives the 3-way valve and I do not need anything more.
The heatingcircuit plugin "bridges" input temperatures to "drive" the outputs, all belonging to external hardware. I wish there was a more detailed documentation.

Besides, you have a very cool DIY weather compensation setup.

Issue closed.
